### PR TITLE
Plugins: Enable plugin runtime install/uninstall capabilities

### PR DESCRIPTION
**Pull request by wbrowne**
**What this PR does / why we need it**:
Enables backend functionality for the bundled marketplace app for managing plugins on the fly.

Adds new HTTP endpoints to expose this functionality (only accessible by admins):

**Installing**

```
curl --request POST \
  --url http://localhost:3000/api/plugins/grafana-github-datasource/install \
  --header 'Content-Type: application/json' \
  --data '{"version": "1.0.10"}'
```
OR
```
curl --request POST \
  --url http://localhost:3000/api/plugins/grafana-github-datasource/install \
  --header 'Content-Type: application/json' 
```

NOTE:
- Installing will remove existing plugin installations if the requested version is different
- The version param is optional. If omitted, it will just fetch the latest

**Uninstalling**

```
curl --request POST \
  --url http://localhost:3000/api/plugins/grafana-github-datasource/uninstall
```

PS: For those interested - Plugin Manager refactoring is ongoing but separate from this work

### CODE REVIEW

File: pkg/plugins/manager/manager_test.go [hunk 1]
```diff
@@ -106,23 +131,66 @@ func TestPluginManager_Init(t *testing.T) {
 	})
 
 	t.Run("With external back-end plugin with valid v2 signature", func(t *testing.T) {
+		pluginsDir := "testdata/valid-v2-signature"
+		pluginFolder := pluginsDir + "/plugin"
```
Comment (aknuds1) [inline pkg/plugins/manager/manager_test.go:135]: Nit, but these are better represented as consts:
```suggestion
		const pluginsDir = "testdata/valid-v2-signature"
		const pluginFolder = pluginsDir + "/plugin"
```

File: pkg/plugins/manager/manager_test.go [hunk 2]
```diff
@@ -106,23 +131,66 @@ func TestPluginManager_Init(t *testing.T) {
...
+			assert.NotNil(t, pm.plugins[pluginID])
+			assert.Equal(t, "datasource", pm.plugins[pluginID].Type)
+			assert.Equal(t, "Test", pm.plugins[pluginID].Name)
+			assert.Equal(t, pluginID, pm.plugins[pluginID].Id)
+			assert.Equal(t, "1.0.0", pm.plugins[pluginID].Info.Version)
+			assert.Equal(t, plugins.PluginSignatureValid, pm.plugins[pluginID].Signature)
+			assert.Equal(t, plugins.GrafanaType, pm.plugins[pluginID].SignatureType)
+			assert.Equal(t, "Grafana Labs", pm.plugins[pluginID].SignatureOrg)
+			assert.Equal(t, pluginFolder, pm.plugins[pluginID].PluginDir)
+			assert.False(t, pm.plugins[pluginID].IsCorePlugin)
```
Comment (aknuds1) [inline pkg/plugins/manager/manager_test.go:168]: I would recommend using go-cmp here, for a more comprehensive equality assertion.

File: pkg/plugins/manager/manager_test.go [hunk 3]
```diff
@@ -221,17 +289,123 @@ func TestPluginManager_IsBackendOnlyPlugin(t *testing.T) {
...
+		assert.NotNil(t, pm.plugins[pluginID])
+		assert.Equal(t, "datasource", pm.plugins[pluginID].Type)
+		assert.Equal(t, "Test", pm.plugins[pluginID].Name)
+		assert.Equal(t, pluginID, pm.plugins[pluginID].Id)
+		assert.Equal(t, "1.0.0", pm.plugins[pluginID].Info.Version)
+		assert.Equal(t, plugins.PluginSignatureValid, pm.plugins[pluginID].Signature)
+		assert.Equal(t, plugins.GrafanaType, pm.plugins[pluginID].SignatureType)
+		assert.Equal(t, "Grafana Labs", pm.plugins[pluginID].SignatureOrg)
+		assert.Equal(t, pluginFolder, pm.plugins[pluginID].PluginDir)
+		assert.False(t, pm.plugins[pluginID].IsCorePlugin)
```
Comment (aknuds1) [inline pkg/plugins/manager/manager_test.go:336]: I would recommend using go-cmp here, for a more comprehensive equality assertion.

File: pkg/plugins/models.go [hunk 4]
```diff
@@ -18,16 +18,16 @@ type PluginNotFoundError struct {
 }
 
 func (e PluginNotFoundError) Error() string {
-	return fmt.Sprintf("plugin with ID %q not found", e.PluginID)
+	return fmt.Sprintf("plugin with ID %s not found", e.PluginID)
```
Comment (aknuds1) [inline pkg/plugins/models.go:21]: Why the change? I think quoting here is better, because the plugin ID could be empty (typical in the case of a bug) or have spaces, in which case an error string that doesn't quote the ID makes for more difficult debugging.

File: pkg/plugins/models.go [hunk 5]
```diff
@@ -18,16 +18,16 @@ type PluginNotFoundError struct {
...
 type DuplicatePluginError struct {
-	Plugin         *PluginBase
-	ExistingPlugin *PluginBase
+	PluginID          string
+	ExistingPluginDir string
 }
 
 func (e DuplicatePluginError) Error() string {
-	return fmt.Sprintf("plugin with ID %q already loaded from %q", e.Plugin.Id, e.ExistingPlugin.PluginDir)
+	return fmt.Sprintf("plugin with ID %s already exists in %s", e.PluginID, e.ExistingPluginDir)
```
Comment (aknuds1) [inline pkg/plugins/models.go:30]: Same comment as before, I think changing to unquoted strings is a change for the worse, since it makes debugging potentially more difficult.

File: pkg/tests/api/plugins/api_install_test.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+	"github.com/grafana/grafana/pkg/bus"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+	"github.com/grafana/grafana/pkg/tests/testinfra"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+var (
```
Comment (aknuds1) [inline pkg/tests/api/plugins/api_install_test.go:20]: These are better represented as consts, since they shouldn't change:
```suggestion
const (
```

File: pkg/tests/api/plugins/api_install_test.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+	require.NoError(t, createUser(t, store, usernameNonAdmin, defaultPassword, false))
+	require.NoError(t, createUser(t, store, usernameAdmin, defaultPassword, true))
+
+	t.Run("Cannot install or uninstall plugin if request is not from an admin", func(t *testing.T) {
+		statusCode, body, err := makePOSTRequest(t, grafanaAPIURL(usernameNonAdmin, grafanaListedAddr, "plugins/grafana-plugin/install"))
+		require.NoError(t, err)
+
+		assert.Equal(t, 403, statusCode)
+		assert.Equal(t, "{\n  \"message\": \"Permission denied\"\n}", body)
```
Comment (aknuds1) [inline pkg/tests/api/plugins/api_install_test.go:40]:
```suggestion
		assert.JSONEq(t, "{\n  \"message\": \"Permission denied\"\n}", body)
```

File: pkg/tests/api/plugins/api_install_test.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+		statusCode, body, err = makePOSTRequest(t, grafanaAPIURL(usernameNonAdmin, grafanaListedAddr, "plugins/grafana-plugin/uninstall"))
+		require.NoError(t, err)
+
+		assert.Equal(t, 403, statusCode)
+		assert.Equal(t, "{\n  \"message\": \"Permission denied\"\n}", body)
```
Comment (aknuds1) [inline pkg/tests/api/plugins/api_install_test.go:46]:
```suggestion
		assert.JSONEq(t, "{\n  \"message\": \"Permission denied\"\n}", body)
```

File: pkg/tests/api/plugins/api_install_test.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+	cmd := models.CreateUserCommand{
+		Login:    username,
+		Password: password,
+		IsAdmin:  isAdmin,
+	}
+	_, err := store.CreateUser(context.Background(), cmd)
+	return err
+}
+
+func makePOSTRequest(t *testing.T, URL string) (int, string, error) {
```
Comment (aknuds1) [inline pkg/tests/api/plugins/api_install_test.go:70]: Post isn't an initialism, so best not written in all uppercase (also, a test helper should check against errors internally, instead of returning them):
```suggestion
func makePostRequest(t *testing.T, URL string) (int, string) {
```

File: pkg/tests/api/plugins/api_install_test.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+	resp, err := http.Post(URL, "application/json", bytes.NewBufferString(""))
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		err := resp.Body.Close()
+		require.NoError(t, err)
```
Comment (aknuds1) [inline pkg/tests/api/plugins/api_install_test.go:75]: There shouldn't be any need to fail the test if freeing a resource failed:
```suggestion
		_ resp.Body.Close()
```

File: pkg/tests/api/plugins/api_install_test.go [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+		IsAdmin:  isAdmin,
+	}
+	_, err := store.CreateUser(context.Background(), cmd)
+	return err
+}
+
+func makePOSTRequest(t *testing.T, URL string) (int, string, error) {
+	resp, err := http.Post(URL, "application/json", bytes.NewBufferString(""))
+	require.NoError(t, err)
+	t.Cleanup(func() {
```
Comment (aknuds1) [inline pkg/tests/api/plugins/api_install_test.go:73]:
```suggestion
	t.Helper()
	
	resp, err := http.Post(URL, "application/json", bytes.NewBufferString(""))
	require.NoError(t, err)
	t.Cleanup(func() {
```

File: pkg/tests/api/plugins/api_install_test.go [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+		err := resp.Body.Close()
+		require.NoError(t, err)
+	})
+	b, err := ioutil.ReadAll(resp.Body)
+
+	return resp.StatusCode, string(b), err
```
Comment (aknuds1) [inline pkg/tests/api/plugins/api_install_test.go:79]:
```suggestion
	b, err := ioutil.ReadAll(resp.Body)
	require.NoError(t, err)

	return resp.StatusCode, string(b)
```

File: pkg/tests/api/plugins/api_install_test.go [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+func createUser(t *testing.T, store *sqlstore.SQLStore, username, password string, isAdmin bool) error {
+	t.Helper()
+
+	cmd := models.CreateUserCommand{
+		Login:    username,
+		Password: password,
+		IsAdmin:  isAdmin,
+	}
+	_, err := store.CreateUser(context.Background(), cmd)
+	return err
```
Comment (aknuds1) [inline pkg/tests/api/plugins/api_install_test.go:67]:
```suggestion
func createUser(t *testing.T, store *sqlstore.SQLStore, username, password string, isAdmin bool) {
	t.Helper()

	cmd := models.CreateUserCommand{
		Login:    username,
		Password: password,
		IsAdmin:  isAdmin,
	}
	_, err := store.CreateUser(context.Background(), cmd)
	require.NoError(t, err)
```

**Review by aknuds1 — COMMENTED** [review]
Generally looks good AFAICT.

File: pkg/plugins/models.go (hunk 4 above)
Comment (wbrowne) [reply pkg/plugins/models.go:21]: I changed this as the quotes end up getting escaped in the logs so it makes it harder to read. Maybe single quoting is the better option in this case?

File: pkg/api/plugins.go [hunk 14]
```diff
@@ -370,6 +370,43 @@ func (hs *HTTPServer) GetPluginErrorsList(_ *models.ReqContext) response.Respons
...
+func (hs *HTTPServer) InstallPlugin(c *models.ReqContext, dto dtos.InstallPluginCommand) response.Response {
+	if !hs.Cfg.MarketplaceAppEnabled {
+		return response.JSON(404, []byte{})
+	}
+	pluginID := c.Params("pluginId")
+
+	err := hs.PluginManager.Install(c.Req.Context(), pluginID, dto.Version)
+	if err != nil {
+		var dupeErr plugins.DuplicatePluginError
+		if errors.As(err, &dupeErr) {
```
Comment (FZambia) [inline pkg/api/plugins.go:382]: `errors.Is` should be better if we don't need further details

File: pkg/api/plugins.go [hunk 15]
```diff
@@ -370,6 +370,43 @@ func (hs *HTTPServer) GetPluginErrorsList(_ *models.ReqContext) response.Respons
...
+			return response.Error(409, "Plugin already installed", err)
+		}
+
+		return response.Error(500, "Failed to install plugin", err)
```
Comment (FZambia) [inline pkg/api/plugins.go:386]: very minor - but better to use codes from http package.

File: pkg/api/plugins.go [hunk 16]
```diff
@@ -370,6 +370,43 @@ func (hs *HTTPServer) GetPluginErrorsList(_ *models.ReqContext) response.Respons
...
+	}
+
+	return response.JSON(200, []byte{})
+}
+
+func (hs *HTTPServer) UninstallPlugin(c *models.ReqContext) response.Response {
+	if !hs.Cfg.MarketplaceAppEnabled {
```
Comment (FZambia) [inline pkg/api/plugins.go:393]: Maybe better to put this check where routes registered?

File: pkg/plugins/backendplugin/manager/manager.go [hunk 17]
```diff
@@ -93,11 +92,51 @@ func (m *manager) Register(pluginID string, factory backendplugin.PluginFactoryF
...
+
+	delete(m.plugins, pluginID)
+
+	m.logger.Debug("Backend plugin unregistered", "pluginId", pluginID)
+	return nil
+}
+
+func (m *manager) Registered(pluginID string) bool {
+	m.pluginsMu.RLock()
+	_, exists := m.plugins[pluginID]
```
Comment (FZambia) [inline pkg/plugins/backendplugin/manager/manager.go:129]: Should we check `decomissioned` flag here too?

File: pkg/plugins/manager/manager.go [hunk 18]
```diff
@@ -228,11 +274,29 @@ func (pm *PluginManager) Apps() []*plugins.AppPlugin {
 	return rslt
 }
 
+func (pm *PluginManager) Panels() []*plugins.PanelPlugin {
+	pm.pluginsMu.RLock()
+	defer pm.pluginsMu.RUnlock()
+
+	var rslt []*plugins.PanelPlugin
```
Comment (FZambia) [inline pkg/plugins/manager/manager.go:281]: can use `copy` here

File: pkg/plugins/manager/manager.go [hunk 19]
```diff
@@ -683,3 +766,90 @@ func (pm *PluginManager) GetDataPlugin(id string) plugins.DataPlugin {
...
+	if plugin != nil {
+		if plugin.Info.Version == version {
+			return plugins.DuplicatePluginError{
+				PluginID:          pluginID,
+				ExistingPluginDir: plugin.PluginDir,
+			}
+		}
+
+		// remove existing installation of plugin
+		err := pm.Uninstall(context.Background(), plugin.Id)
```
Comment (FZambia) [inline pkg/plugins/manager/manager.go:781]: As far as I know there are several plugin directories checked on start (may be wrong), is it ok if plugin removed from one dir but have been downloaded to another?

File: pkg/plugins/manager/manager.go [hunk 20]
```diff
@@ -683,3 +766,90 @@ func (pm *PluginManager) GetDataPlugin(id string) plugins.DataPlugin {
...
+
+	return nil
+}
+
+func (pm *PluginManager) removeStaticRoute(pluginID string) {
+	routes := pm.staticRoutes
+
+	for i, route := range routes {
+		if pluginID == route.PluginId {
+			r := make([]*plugins.PluginStaticRoute, 0)
```
Comment (FZambia) [inline pkg/plugins/manager/manager.go:848]:
```go
pm.staticRoutes = append(routes[:i], routes[i+1:]...)
return
```

**Review by FZambia — COMMENTED** [review]
Left some comments: minor tweaks and some questions

File: pkg/plugins/manager/manager.go (hunk 18 above)
Comment (wbrowne) [reply pkg/plugins/manager/manager.go:281]: `pm.panels` is a map so unfortunately `copy` can't give us any help in this case. However, I will pre-allocate those slices now that I see that

**Review by achatterjee-grafana — CHANGES_REQUESTED** [review]
Added doc change request.

File: docs/sources/administration/configuration.md [hunk 21]
```diff
@@ -1446,6 +1446,12 @@ Enter a comma-separated list of plugin identifiers to identify plugins that are
 
 Custom install/learn more url for enterprise plugins. Defaults to https://grafana.com/grafana/plugins/.
 
+### marketplace_app_enabled
+
+> **Note:** Available in Grafana 8.0 and later versions.
+
+Available to Grafana administrators only, the plugin marketplace app is set to `false` by default. Set it to `true` if to enable the app.
```
Comment (achatterjee-grafana) [inline docs/sources/administration/configuration.md:1453]:
```suggestion
Available to Grafana administrators only, the plugin marketplace app is set to `false` by default. Set it to `true` to enable the app.
```

Comment (achatterjee-grafana) [reply docs/sources/administration/configuration.md:1453]: Sorry for the typo!

**Review by achatterjee-grafana — CHANGES_REQUESTED** [review]

File: docs/sources/administration/configuration.md (hunk 21 above)
Comment (wbrowne) [reply docs/sources/administration/configuration.md:1453]: Missed it too 😆  Thanks!

**Review by achatterjee-grafana — APPROVED** [review]
Doc change looks good.

**Comment by marcusolsson** [comment]
I don't have a strong opinion on where the documentation should live, but the first place I come to think of is a dedicated page under the top-level [Plugins](https://grafana.com/docs/grafana/latest/plugins/). Alternatively, we can add on the [Install plugins](https://grafana.com/docs/grafana/latest/plugins/installation/) page? WDYT?

File: docs/sources/plugins/marketplace.md [hunk 22]
```diff
@@ -0,0 +1,12 @@
++++
+title = "Plugin marketplace"
+aliases = ["/docs/grafana/latest/plugins/marketplace/"]
+weight = 1
++++
+
+Marketplace for Grafana is a plugin bundled with Grafana versions 8.0+. It allows users to browse and manage plugins from within Grafana. Access is limited to Grafana Admins.
```
Comment (marcusolsson) [inline docs/sources/plugins/marketplace.md:7]: > Access is limited to Grafana Admins

Access to what? The entire app or installing/uninstalling? Can Viewers at least browse plugins?

File: docs/sources/plugins/marketplace.md [hunk 23]
```diff
@@ -0,0 +1,12 @@
...
+
+[screenshot placeholder]
+
+In order to use the Marketplace for Grafana, it must be enabled in the Grafana [configuration]({{< relref "../administration/configuration.md#marketplace_app_enabled" >}}).
```
Comment (marcusolsson) [inline docs/sources/plugins/marketplace.md:11]:
```suggestion
To use the Marketplace for Grafana, you first need to enable it in the Grafana [configuration]({{< relref "../administration/configuration.md#marketplace_app_enabled" >}}).
```

**Comment by wbrowne** [comment]
Good point @marcusolsson. I have just added placeholder for a dedicated page under the plugins section for now. I think it's worth mentioning it from within the `Install plugin on local Grafana` section of the install plugins page

**Review by marcusolsson — COMMENTED** [review]
I'm thinking we should we add tasks for installing and uninstalling plugins?

## Install a plugin from the Marketplace

To install a plugin ...

## Uninstall a plugin from the Marketplace

To uninstall a plugin ...

**Review by marefr — COMMENTED** [review]
Looks great 👍 Great work. Added some comments, nits and questions. Planning to take this for a spin tomorrow

Disclaimer: Used GitHub app for ios reviewing this so might have missed a few things :)

File: pkg/api/plugins.go [hunk 24]
```diff
@@ -370,6 +370,37 @@ func (hs *HTTPServer) GetPluginErrorsList(_ *models.ReqContext) response.Respons
...
 
+func (hs *HTTPServer) InstallPlugin(c *models.ReqContext, dto dtos.InstallPluginCommand) response.Response {
+	pluginID := c.Params("pluginId")
+
+	err := hs.PluginManager.Install(c.Req.Context(), pluginID, dto.Version)
+	if err != nil {
+		var dupeErr plugins.DuplicatePluginError
+		if errors.As(err, &dupeErr) {
+			return response.Error(http.StatusConflict, "Plugin already installed", err)
+		}
```
[the remark this answers is not in the document]
Comment (wbrowne) [reply pkg/api/plugins.go:381]: Will make a change to propagate the not found error to the api level 👍 

The marketplace has done this previously, but we haven't come to a full conclusion yet I don't think. I'll also propagate this error as that should at least provide some feedback to the user (with suggested compatible version)

File: pkg/cmd/grafana-server/main.go [hunk 25]
```diff
@@ -38,7 +38,7 @@ import (
 )
 
 // The following variables cannot be constants, since they can be overridden through the -X link flag
-var version = "5.0.0"
+var version = "7.5.0"
```
[the remark this answers is not in the document]
Comment (wbrowne) [reply pkg/cmd/grafana-server/main.go:41]: Because I was running the backend directly without any wrappers, this was the default value for the `cfg.BuildVersion`, which is used to check compatibility against the gcom API. I can probably revert, but I didn't think there's any harm?

File: pkg/plugins/backendplugin/grpcplugin/grpc_plugin.go [hunk 26]
```diff
@@ -100,6 +101,19 @@ func (p *grpcPlugin) Exited() bool {
 	return true
 }
 
+func (p *grpcPlugin) Decommission() error {
+	p.mutex.RLock()
```
[the remark this answers is not in the document]
Comment (wbrowne) [reply pkg/plugins/backendplugin/grpcplugin/grpc_plugin.go:105]: Good eye 🥇

File: pkg/plugins/backendplugin/ifaces.go [hunk 27]
```diff
@@ -10,8 +10,12 @@ import (
 
 // Manager manages backend plugins.
 type Manager interface {
-	// Register registers a backend plugin
-	Register(pluginID string, factory PluginFactoryFunc) error
+	// RegisterAndStart registers and starts a backend plugin
+	RegisterAndStart(ctx context.Context, pluginID string, factory PluginFactoryFunc) error
+	// UnregisterAndStop unregisters and stops a backend plugin
+	UnregisterAndStop(ctx context.Context, pluginID string) error
```
[the remark this answers is not in the document]
Comment (wbrowne) [reply pkg/plugins/backendplugin/ifaces.go:16]: I think from a high-level API perspective it's clearer that these two functions are related, where one starts the thing, and another ends it. Decommission is a bit leaky in this case.

Alternatively, we could even just simplify to `Start` and `Stop`, but that would probably be confusing when we have `StartPlugin` for the renderer. Unless that then becomes `StartRenderer`... Down the 🐰 hole we go

File: pkg/plugins/backendplugin/manager/manager.go [hunk 28]
```diff
@@ -93,11 +92,51 @@ func (m *manager) Register(pluginID string, factory backendplugin.PluginFactoryF
 
 	m.plugins[pluginID] = plugin
 	m.logger.Debug("Backend plugin registered", "pluginId", pluginID)
+
+	m.start(ctx, plugin)
```
[the remark this answers is not in the document]
Comment (wbrowne) [reply pkg/plugins/backendplugin/manager/manager.go:96]: Yeah is good to be aware of at least. I also don't think it's makes a difference, but something to keep an eye on!

File: pkg/plugins/manager/manager.go [hunk 29]
```diff
@@ -683,3 +766,85 @@ func (pm *PluginManager) GetDataPlugin(id string) plugins.DataPlugin {
...
+	return nil
+}
+
+func (pm *PluginManager) Uninstall(ctx context.Context, pluginID string) error {
+	plugin := pm.GetPlugin(pluginID)
+	if plugin == nil {
+		return plugins.PluginNotFoundError{PluginID: pluginID}
+	}
+
+	if pm.BackendPluginManager.Registered(pluginID) {
```
[the remark this answers is not in the document]
Comment (wbrowne) [reply pkg/plugins/manager/manager.go:806]: We do some checks but can improve this!

File: pkg/cmd/grafana-server/main.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (marefr) [reply pkg/cmd/grafana-server/main.go:41]: I don't think there should be any harm :smile: 

But don't you see this when starting Grafana `INFO[05-12|10:34:18] Starting Grafana                         logger=server version=7.5.0-pre commit=4f03c7c9e6 branch=feat_httpclientprovider compiled=2021-05-11T20:14:22+0200     `? Or what do you mean without wrappers?

File: pkg/plugins/backendplugin/ifaces.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (marefr) [reply pkg/plugins/backendplugin/ifaces.go:16]: No biggie. Was just curious. Lets skip this for now

File: pkg/cmd/grafana-server/main.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (wbrowne) [reply pkg/cmd/grafana-server/main.go:41]: Nah I'm special 😇 
`INFO[05-12|11:18:10] Starting Grafana logger=server version=5.0.0 commit=NA branch=main compiled=2021-05-12T11:18:10+0200`

It's because I build and run grafana-server directly, I don't use `make run` or anything so it doesn't go through the regular flow of setting the version from package.json 🙈

[the remark this answers is not in the document]
Comment (marefr) [reply pkg/cmd/grafana-server/main.go:41]: I see. That's your own fault :smile: This version would potentially be out-of-date for you soon again so continuing updating this is not sustainable but would work for now.

File: docs/sources/plugins/marketplace.md [hunk 30]
```diff
@@ -0,0 +1,20 @@
++++
+title = "Plugin marketplace"
+aliases = ["/docs/grafana/latest/plugins/marketplace/"]
+weight = 1
++++
+
```
[the remark this answers is not in the document]
Comment (wbrowne) [reply docs/sources/plugins/marketplace.md:6]: Do you mean that `docs/sources/plugins/marketplace.md` should match the title here
```
+++
title = "Plugin marketplace"
...
```

IE `docs/sources/plugins/plugin-marketplace.md`◀️ ▶️ `title = "Plugin marketplace"`?

**Review by FZambia — APPROVED** [review]
👍

File: docs/sources/plugins/marketplace.md [hunk 31]
```diff
@@ -0,0 +1,20 @@
...
+Marketplace for Grafana is a plugin bundled with Grafana versions 8.0+. It allows users to browse and manage plugins from within Grafana. Only Grafana Admins can access and use the Marketplace.
+
+[screenshot placeholder]
+
+To use the Marketplace for Grafana, you first need to enable it in the Grafana [configuration]({{< relref "../administration/configuration.md#marketplace_app_enabled" >}}).
+
+## Install a plugin from the Marketplace
```
[the remark this answers is not in the document]
Comment (wbrowne) [reply docs/sources/plugins/marketplace.md:15]: @oddlittlebird Are the docs fine to merge as is, and we can fill in the gaps during the beta?

File: pkg/api/api.go [hunk 32]
```diff
@@ -281,7 +281,14 @@ func (hs *HTTPServer) registerRoutes() {
 		apiRoute.Get("/plugins/:pluginId/health", routing.Wrap(hs.CheckHealth))
 		apiRoute.Any("/plugins/:pluginId/resources", hs.CallResource)
 		apiRoute.Any("/plugins/:pluginId/resources/*", hs.CallResource)
-		apiRoute.Any("/plugins/errors", routing.Wrap(hs.GetPluginErrorsList))
+		apiRoute.Get("/plugins/errors", routing.Wrap(hs.GetPluginErrorsList))
+
+		if hs.Cfg.MarketplaceAppEnabled {
+			apiRoute.Group("/plugins", func(pluginRoute routing.RouteRegister) {
+				pluginRoute.Post("/:pluginId/install", bind(dtos.InstallPluginCommand{}), routing.Wrap(hs.InstallPlugin))
```
[the remark this answers is not in the document]
Comment (wbrowne) [reply pkg/api/api.go:289]: Will address this during the beta if that's okay with you

[the remark this answers is not in the document]
Comment (marefr) [reply pkg/api/api.go:289]: Sure

**Review by marefr — APPROVED** [review]
LGTM

Still haven't had time to try it out. Would do that during the beta I suppose. Great work

File: docs/sources/plugins/marketplace.md (hunk 30 above)
[the remark this answers is not in the document]
Comment (oddlittlebird) [reply docs/sources/plugins/marketplace.md:6]: Yes.

File: docs/sources/plugins/marketplace.md (hunk 31 above)
[the remark this answers is not in the document]
Comment (oddlittlebird) [reply docs/sources/plugins/marketplace.md:15]: Yes, that works

**Review by oddlittlebird — APPROVED** [review]
